src/models/db/case.js

```diff
@@ -1,72 +1,134 @@
+    .from(TABLE_NAME)
+    .select()
+    .where('CASE_ID', primaryKey)
+    .then((results) => results.map(convertFromPersistent))
+    .then((results) => {
+      if (results.length === 1) {
+        return results[0]
+      } else if (results.length === 0) {
+        return null
+      } else {
+        assert(false)
       }
-      data.notification_targets = JSON.parse(data.notification_targets)
-      delete data.created_at
-      delete data.updated_at
-      if (!Array.isArray(data.notaries)) {
-        data.notaries = [ data.notaries ]
+    })
+}
+
+function getCaseByUuid (caseUuid, options) {
+  return knexOrTransaction(knex, options)
+    .from(TABLE_NAME)
+    .select()
+    .where('CASE_UUID', caseUuid)
+    .then((results) => results.map(convertFromPersistent))
+    .then((results) => {
+      if (results.length === 1) {
+        return results[0]
+      } else if (results.length === 0) {
+        return null
+      } else {
+        assert(false)
       }
-      return data
-    }
+    })
+}
 
-    static convertToPersistent (data) {
-      data.execution_condition = JSON.stringify(data.execution_condition)
-      data.exec_cond_fulfillment = JSON.stringify(data.exec_cond_fulfillment)
-      data.notification_targets = JSON.stringify(data.notification_targets)
-      data.notaries = data.notaries[0]
-      return data
-    }
+function save (caseObject, options) {
+  const myKnex = knexOrTransaction(knex, options)
+  const persistentCaseObject = convertToPersistent(caseObject)
+  if (persistentCaseObject.CASE_ID) {
+    return myKnex(TABLE_NAME).update(persistentCaseObject)
+      .where('CASE_ID', persistentCaseObject.CASE_ID)
+      .then()
+  } else {
+    return myKnex(TABLE_NAME).insert(persistentCaseObject).then()
+  }
+}
 
-    isFinalized () {
-      return this.state === 'executed' || this.state === 'rejected'
-    }
+function * _upsertCase (caseData, options) {
+  assert(options.transaction)
+  // SQLite's implementation of upsert does not tell you whether it created the
+  // row or whether it already existed. Since we need to know to return the
+  // correct HTTP status code we unfortunately have to do this in two steps.
+  const existingCase = yield getCaseByUuid(caseData.case_uuid, options)
+  if (existingCase) {
+    yield save(_.merge({}, existingCase, caseData), options)
+  } else {
+    yield save(caseData, options)
   }
+  return Boolean(existingCase)
+}
 
-  Case.validateExternal = validator.create('Case')
-  Case.tableName = 'cases'
+function * upsertCase (caseData, options) {
+  if (options && options.transaction) {
+    return yield _upsertCase(caseData, options)
+  } else {
+    let result
+    yield transaction(function * (transaction) {
+      result = yield _upsertCase(caseData,
+        _.assign({}, options || {}, {transaction}))
+    })
+    return result
+  }
+}
 
-  PersistentKnexModelMixin(Case, knex)
+function convertFromPersistent (data) {
+  data = _.cloneDeep(data)
```

Comment:
I started trying to avoid reassignment here and making it more functional, not a big issue though